pkg/firewall/client.go

```diff
@@ -2,66 +2,124 @@ package firewall
 		k8sClient: k8sClient,
 	}
 }
 
+const (
+	ProtocolTCP = "tcp"
+	PortSSH     = "22"
+)
+
 func (c *Client) CreateBastionFirewallRule(ctx context.Context, cluster *capg.GCPCluster) error {
-	// TODO
+	fwName := bastionFirewallPolicyRuleName(cluster.Name)
+	ipProtocol := ProtocolTCP
 
-	return nil
-}
+	tagName := fmt.Sprintf("%s-bastion", cluster.GetName())
 
-func (c *Client) AssignBastionFirewallRule(ctx context.Context, cluster *capg.GCPCluster) error {
-	// TODO
-	/*
-		_, err := c.dnsService.ResourceRecordSets.Create(cluster.Spec.Project, cluster.Name, record).
-			Context(ctx).
-			Do()
+	rule := &computepb.Firewall{
+		Allowed: []*computepb.Allowed{
+			{
+				IPProtocol: &ipProtocol,
+				Ports:      []string{PortSSH},
+			},
+		},
+		Description: proto.String("allow port 22 for SSH to"),
+		Direction:   proto.String(computepb.Firewall_INGRESS.String()),
+		Name:        &fwName,
+		Network:     cluster.Status.Network.SelfLink,
+		TargetTags:  []string{tagName},
+	}
+
+	req := &computepb.InsertFirewallRequest{
+		Project:          cluster.Spec.Project,
+		FirewallResource: rule,
+	}
 
-		if hasHttpCode(err, http.StatusConflict) {
+	op, err := c.fwService.Insert(ctx, req)
+	if err != nil {
+		if isAlreadyExistError(err) {
+			// pass thru, resource already exists
 			return nil
+		} else {
+			return microerror.Mask(err)
 		}
-		return microerror.Mask(err)
+	}
+
+	err = op.Wait(ctx)
+
+	if err != nil {
+		if isAlreadyExistError(err) {
+			// pass thru, resource already exists
+		} else {
+			return microerror.Mask(err)
+		}
+	}
+
+	return nil
+}
+
+func (c *Client) AssignBastionFirewallRule(ctx context.Context, cluster *capg.GCPCluster) error {
 
-	*/
 	return nil
 }
 
 func (c *Client) DisassociateBastionFirewallRule(ctx context.Context, cluster *capg.GCPCluster) error {
-	// TODO
 
 	return nil
 }
 
 func (c *Client) DeleteBastionFirewallRule(ctx context.Context, cluster *capg.GCPCluster) error {
-	// TODO
+	req := &computepb.DeleteFirewallRequest{
+		Project:  cluster.Spec.Project,
+		Firewall: bastionFirewallPolicyRuleName(cluster.Name),
+	}
+
+	op, err := c.fwService.Delete(ctx, req)
+	if isNotFoundError(err) {
+		// pass thru, resource is already deleted
+		return nil
+	} else if err != nil {
+		return microerror.Mask(err)
+	}
+	err = op.Wait(ctx)
+
+	if isNotFoundError(err) {
+		// pass thru, resource is already deleted
+	} else if err != nil {
+		return microerror.Mask(err)
+	}
+
 	return nil
 }
 
-/*
-func hasHttpCode(err error, statusCode int) bool {
-	var googleErr *googleapi.Error
-	if errors.As(err, &googleErr) {
-		if googleErr.Code == statusCode {
-			return true
-		}
-	}
+func bastionFirewallPolicyRuleName(clusterName string) string {
+	return fmt.Sprintf("allow-%s-bastion-ssh", clusterName)
+}
+
+func isAlreadyExistError(err error) bool {
+	return strings.Contains(err.Error(), "already exists")
+}
 
-	return false
+func isNotFoundError(err error) bool {
+	return strings.Contains(err.Error(), "was not found")
 }
```

Comment:
The `sigs.k8s.io/cluster-api-provider-gcp/cloud/gcperrors` package contains some helper functions for this kind of thing.

Reply:
done